gnucash/gnome/assistant-stock-transaction.cpp

```diff
@@ -0,0 +1,1151 @@
+}
+
+static void refresh_page_dividend (GtkWidget *widget, gpointer user_data)
+{
+    return;
+}
+
+static void refresh_page_capgains (GtkWidget *widget, gpointer user_data)
+{
+    return;
+}
+
+static void add_error (StringVec& errors, const std::string& format_str,
+                       const std::string& arg)
+{
+    gchar *buf = g_strdup_printf (format_str.c_str (), arg.c_str());
+    errors.emplace_back (std::string {buf});
+    g_free (buf);
+}
+
+static void
+check_page (GtkListStore *list, gnc_numeric& debit, gnc_numeric& credit,
+            uint splitfield, Account *acct, GtkWidget *memo, GtkWidget *gae,
+            gnc_commodity *comm, bool ignore_account,
+            std::string page, StringVec& errors)
+{
+    if (splitfield == DISABLED)
+        return;
+
+    const gchar* amtstr;
+    gnc_numeric amount;
+    bool debit_side = (splitfield & ENABLED_DEBIT);
+
+    if (gnc_amount_edit_expr_is_valid (GNC_AMOUNT_EDIT (gae), &amount, true, nullptr))
+    {
+        if (splitfield & ALLOW_ZERO)
+            amtstr = "";
+        else
+        {
+            add_error (errors, "Amount for %s is missing", page);
+            amtstr = "(missing)";
+        }
+    }
+    else
+    {
+        if (!(splitfield & ALLOW_NEGATIVE))
+        {
+            if ((splitfield & ALLOW_ZERO) && gnc_numeric_negative_p (amount))
+                add_error (errors, "Amount for %s must not be negative.", page);
+            else if (!(splitfield & ALLOW_ZERO) && !gnc_numeric_positive_p (amount))
+                add_error (errors, "Amount for %s must be positive.", page);
+        }
+        if (gnc_numeric_negative_p (amount))
+        {
+            amount = gnc_numeric_neg (amount);
+            debit_side = !debit_side;
+        }
+        if (debit_side)
+            debit = gnc_numeric_add_fixed (debit, amount);
+        else
+            credit = gnc_numeric_add_fixed (credit, amount);
+        amtstr = xaccPrintAmount (amount, gnc_commodity_print_info (comm, true));
+    }
+
+    auto memostr = gtk_entry_get_text (GTK_ENTRY (memo));
+    const gchar *acctstr;
+
+    if (ignore_account)
+        acctstr = "";
+    else if (acct)
+        acctstr = xaccAccountGetName (acct);
+    else
+    {
+        add_error (errors, "Account for %s is missing", page);
+        acctstr = "(missing)";
+    }
+
+    GtkTreeIter iter;
+    gtk_list_store_append (list, &iter);
+    gtk_list_store_set (list, &iter,
+                        SPLIT_COL_ACCOUNT, acctstr,
+                        SPLIT_COL_MEMO, memostr,
+                        SPLIT_COL_DEBIT, debit_side ? amtstr : "",
+                        SPLIT_COL_CREDIT, !debit_side ? amtstr : "",
+                        -1);
+}
+
+static std::string join(const StringVec& v, char c)
+{
+    std::string s;
+    for (StringVec::const_iterator p = v.begin(); p != v.end(); ++p)
+    {
+        s += *p;
+        if (p != v.end() - 1)
+            s += c;
+    }
```

Comment:
This (std::for_each) won't work because its 3rd arg where you've offered the lambda does not get the iterator as an argument. It gets the vector element itself. https://en.cppreference.com/w/cpp/algorithm/for_each

Reply:
> This (std::for_each) won't work because its 3rd arg where you've offered the lambda does not get the iterator as an argument. It gets the vector element itself. https://en.cppreference.com/w/cpp/algorithm/for_each

Having said that^, this function `join` amended to use std::for_each which doesn't need the iterator.